### ISSUE TITLE
Nested Form Parameters

### ISSUE THREAD

**Issue by alihalabyah** [teamon/tesla#184]
Hi Guys,

Can you please provide an example showing how to accomplish the below CURL:

```
curl https://api.elixir.com/v2/objects \
  -H 'Authorization: Token token=test_-ZbfdrY8zCdUW1nC8zEDvy5aGrzWwEvq' \
  -d 'title=Mr' \
  -d 'first_name=John' \
  -d 'last_name=Smith' \
  -d 'email=johnsmith@fakegmail.com' \
  -d 'gender=male' \
  -d 'dob=1980-01-22' \
  -d 'country=GBR' \
  -d 'addresses[][building_number]=100' \
  -d 'addresses[][street]=Main Street' \
  -d 'addresses[][town]=London' \
  -d 'addresses[][postcode]=SW4 6EH' \
  -d 'addresses[][country]=GBR'
```

I'm using Phoenix and sending receiving the request in the below structure:

```
{
  "title": "Mr",
  "last_name": "Doe",
  "gender": "male",
  "first_name": "John",
  "email": "john.doe@doe.com",
  "dob": "1980-01-22",
  "country": "GBR",
  "addresses": [
    {
      "town": "London",
      "street": "Main Street",
      "postcode": "SW4 6EH",
      "country": "GBR",
      "building_number": "100"
    }
  ]
}
```

and Phoenix is converting it into a map and from there I'm struggling in passing correct params.

Any idea how this can be fixed?

**Comment by amatalai** [comment]
I think that what you want to achieve should look something like this*
```elixir
defmodule MyApp.API do
  use Tesla

  plug Tesla.Middleware.Headers, %{"Authorization" => "Token token=test_-ZbfdrY8zCdUW1nC8zEDvy5aGrzWwEvq"}
  plug Tesla.Middlware.JSON

  def send do
    body = %{
      "addresses" => [
        %{
          "building_number" => "100",
          "street" => "Main Street",
          "town" => "London",
          "postcode" => "SW4 6EH",
          "country" => "GBR"
        }
      ],
      "country" => "GBR",
      "dob" => "1980-01-22",
      "email" => "johnsmith@fakegmail.com",
      "first_name" => "John",
      "gender" => "male",
      "last_name" => "Smith",
      "title" => "Mr"
    }

    post("https://api.elixir.com/v2/objects", body)
  end
end

MyApp.API.send()
```

\* I didn't test it in any way, it may contain typos or bugs

**Comment by teamon** [comment]
@alihalabyah I'm not sure I fully understand what are you asking for - tesla has nothing to do with phoenix. I can only guess that if you want to mimic the curl request you will need to use `Tesla.Middleware.FormUrlencoded` middleware.

**Comment by teamon** [comment]
@alihalabyah I'm closing this issue due to inactivity. Feel free to comment/reopen if needed.

**Comment by chulkilee** [comment]
I don't think there is "official" way for two-way conversion between url encoded form and nested structure. So I think this should be done by own middleware

**Comment by Fire-Dragon-DoL** [comment]
@teamon I stumbled upon this too. It doesn't support nested maps. A pretty common usecase is when dealing with Stripe API. They request you to provide `metadata[custom_arg]=something` for metadata values. I tried using `body: %{description: "...", metadata: %{"custom_arg" => 123}}` but it raised an exception.

I also found a potential solution. Replace (in FormUrlencoded middleware) `URI.encode_query` and `URI.decode_query` with `Plug.Conn.Query.encode` and `Plug.Conn.Query.decode`. The `Plug` version supports nested maps.

My recommendation is either provide an option to use custom encoder/decoder for the middleware, or accept `Plug` as an optional dependency to allow decoding through it.

Alternatively, another middleware that works in a nested fashion too.

If it could be of any interest, I'm open to writing a PR for it.
